File: 240923/미로 타워 디펜스/maze-tower-defense.py
def initialize():
    global monsters
    x, y = n // 2, n // 2
    idx = 0
    dist = 1
    direction = 0
    move = 0
    while True:
        for _ in range(dist):
            monsters[idx] = grid[x][y]
            pos2idx[(x, y)] = idx
            nx = x + maze_dirs[direction][0]
            ny = y + maze_dirs[direction][1]
            if (nx, ny) == (0, -1):
                return
            idx += 1
            x, y = nx, ny
        move += 1
        direction = (direction + 1) % 4
        if move == 2:
            dist += 1
            move = 0


def attack(d, p):
    global monsters, score
    x, y = n // 2, n // 2
    dx, dy = attack_dirs[d]
    for i in range(1, p + 1):
        nx = x + dx * i
        ny = y + dy * i
        score += monsters[pos2idx[(nx, ny)]]
        monsters[pos2idx[(nx, ny)]] = -1


def remove_repeated_monsters():
    global score
    flag = False
    target = 0
    cnt = 0
    for i in range(n ** 2):
        if monsters[i] == monsters[target]:
            cnt += 1
        else:
            if 4 <= cnt:
                flag = True
                score += monsters[target] * cnt
                for j in range(target, i):
                    monsters[j] = -1
            target = i
            cnt = 1
    return flag


def tide_up():
    global monsters
    removed_cnt = monsters.count(-1)
    monsters = [monster for monster in monsters if monster != -1] + [0] * removed_cnt


def pair_monster_nums():
    global monsters
    new_monsters = [0]
    group = []
    for i in range(1, n ** 2):
        if not group:
            group.append(monsters[i])
        elif monsters[i] == group[0]:
            group.append(monsters[i])
        else:
            new_monsters.append(len(group))
            new_monsters.append(group[0])
            group = [monsters[i]]
    monsters = [0] * (n ** 2)
    for i in range(len(new_monsters)):
        if i >= (n ** 2):
            break
        monsters[i] = new_monsters[i]


n, m = map(int, input().split())
grid = [list(map(int, input().split())) for _ in range(n)]
monsters = [-1] * (n ** 2)
pos2idx = {}
attack_dirs = ((0, 1), (1, 0), (0, -1), (-1, 0))  # → ↓ ← ↑
maze_dirs = ((0, -1), (1, 0), (0, 1), (-1, 0))
initialize()
score = 0
for _ in range(m):
    d, p = map(int, input().split())
    attack(d, p)
    tide_up()
    while True:
        if not remove_repeated_monsters():
            break
        tide_up()
    pair_monster_nums()
print(score)

File: 240923/미로 타워 디펜스/maze-tower-defense.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {

	static BufferedReader br;
	static StringBuilder sb;
	static StringTokenizer st;

	static final int[][] attackDirs = { { 0, 1 }, { 1, 0 }, { 0, -1 }, { -1, 0 } }; // → ↓ ← ↑
	static final int[][] mazeDirs = { { 0, -1 }, { 1, 0 }, { 0, 1 }, { -1, 0 } };

	static int n, m;
	static int[][] grid;
	static int[] monsters;
	static Map<Pos, Integer> posIndexMap;
	static int score;

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
		sb = new StringBuilder();
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		grid = new int[n][n];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < n; j++) {
				grid[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		monsters = new int[n * n];
		Arrays.fill(monsters, -1);
		posIndexMap = new HashMap<>();
		score = 0;
		initialize();
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());
			final int d = Integer.parseInt(st.nextToken());
			final int p = Integer.parseInt(st.nextToken());
			attack(d, p);
			tideUp();
			while (true) {
				if (!removeRepeatedMonsters()) {
					break;
				}
				tideUp();
			}
			pairMonsterNums();
		}
		System.out.println(score);
	}

	private static void initialize() {
		int x = n / 2;
		int y = n / 2;
		int index = 0;
		int dist = 1;
		int dir = 0;
		int move = 0;
		while (true) {
			for (int i = 0; i < dist; i++) {
				monsters[index] = grid[x][y];
				posIndexMap.put(new Pos(x, y), index);
				final int nx = x + mazeDirs[dir][0];
				final int ny = y + mazeDirs[dir][1];
				if (nx == 0 && ny == -1) {
					return;
				}
				index++;
				x = nx;
				y = ny;
			}
			move++;
			dir = (dir + 1) % 4;
			if (move == 2) {
				dist++;
				move = 0;
			}
		}
	}

	private static void attack(final int d, final int p) {
		final int x = n / 2;
		final int y = n / 2;
		for (int i = 1; i <= p; i++) {
			final int nx = x + attackDirs[d][0] * i;
			final int ny = y + attackDirs[d][1] * i;
			final int index = posIndexMap.get(new Pos(nx, ny));
			score += monsters[index];
			monsters[index] = -1;
		}
	}

	private static void tideUp() {
		final int[] temp = new int[n * n];
		int index = 0;
		for (int i = 0; i < n * n; i++) {
			if (monsters[i] == -1) {
				continue;
			}
			temp[index++] = monsters[i];
		}
		monsters = temp;
	}

	private static boolean removeRepeatedMonsters() {
		boolean flag = false;
		int target = 0;
		int count = 0;
		for (int i = 0; i < n * n; i++) {
			if (monsters[i] == monsters[target]) {
				count++;
			} else {
				if (count >= 4) {
					flag = true;
					score += monsters[target] * count;
					for (int j = target; j < i; j++) {
						monsters[j] = -1;
					}
				}
				target = i;
				count = 1;
			}
		}
		return flag;
	}

	private static void pairMonsterNums() {
		final int[] newMonsters = new int[n * n];
		final Queue<Integer> group = new ArrayDeque<>();
		int index = 1;
		for (int i = 1; i < n * n; i++) {
			if (group.isEmpty()) {
				group.add(monsters[i]);
			} else if (monsters[i] == group.peek()) {
				group.add(monsters[i]);
			} else {
				if (index == n * n) {
					break;
				}
				newMonsters[index++] = group.size();
				newMonsters[index++] = group.peek();
				group.clear();
				group.add(monsters[i]);
			}
		}
		monsters = newMonsters;
	}

	static class Pos {

		final int x;
		final int y;

		public Pos(final int x, final int y) {
			this.x = x;
			this.y = y;
		}

		@Override
		public boolean equals(Object object) {
			final Pos other = (Pos) object;
			return this.x == other.x && this.y == other.y;
		}

		@Override
		public int hashCode() {
			return 100 * x + y;
		}
	}
}